### DISCUSSION TITLE
MUSLLINUX by default?

### DISCUSSION THREAD

**Discussion by henryiii** [pypa/cibuildwheel#839] [General]
I believe the current musllinux is active by default, which will double linux build times, and possibly not work in quite a few cases (such as builds with dependency installs, etc) right away. I think we should at least clearly document how to turn them off, possibly in the release notes (`skip=*muslinux*` or such). It almost would be nice to have a toggle that's off by default, but I don't want to add extra toggles, and we don't leave PyPy off by default either. It's just a sudden addition that could be surprising. I also don't want to discourage adoption of musllinux packages.

Also, I'm okay with a release before #799 is done, though it will make it easier to use. Given the size of #799, it will likely take a bit of time; getting something out with musllinux quickly would be nice (but see the above worry).

Just providing a place to discuss if anyone has thoughts.

**Comment by henryiii** [comment]
Also, musllinux will likely be much slower than manylinux until upstream packages (like NumPy) start shipping wheels.

**Comment by joerick** [comment]
Thanks for raising this @henryiii. I hadn't fully considered how many projects this could impact during the upgrade. 

> I think we should at least clearly document how to turn them off, possibly in the release notes (skip=\*muslinux\* or such).

Strong agree on this. I'll be sure to add that in the release notes. We should also add a clear example to the CIBW_SKIP docs.

> It almost would be nice to have a toggle that's off by default, but I don't want to add extra toggles, and we don't leave PyPy off by default either.

Yeah, this is worth talking about, but I think I'd agree that we probably don't want to change the current behaviour this time. 

That said, it's not impossible that one day we might add a more exotic platform that we don't want to build by default. Maybe one day there's a niche platform like Pyodide that we'd like to add support for, without enabling by default? Previously we've discussed having a default on CIBW_BUILD, but then when called with `cp310-*` it would include the exotic platform. So then we'd probably be looking at flags like CIBW_ENABLE_EXOTICPLATFORMX=1, or `CIBW_ENABLE=exotic_platform_x`. Which is a bit more hoop-jumping than I'd like, but maybe that will be the right solution one day.

> Also, I'm okay with a release before #799 is done, though it will make it easier to use.

Hmm. Maybe we do an alpha i.e. v2.2.0a1 now so that eager beavers can try out musllinux now, but wait until #799 is ready for the proper release? I agree that it will be really useful to have #799 when we release musl, as that will really help the more complicated projects customise their configs, so they're more likely to adopt musllinux from the start. 

What do you think? How does an alpha sound to you?

**Reply by henryiii** [reply]
An alpha sounds good to me!

**Comment by mayeut** [comment]
>> I think we should at least clearly document how to turn them off, possibly in the release notes (skip=*muslinux* or such).

> Strong agree on this. I'll be sure to add that in the release notes. We should also add a clear example to the CIBW_SKIP docs.

+1 on this.

>> It almost would be nice to have a toggle that's off by default, but I don't want to add extra toggles, and we don't leave PyPy off by default either.

> Yeah, this is worth talking about, but I think I'd agree that we probably don't want to change the current behaviour this time.

I'm not in favor of turning things off by default. Maybe as @joerick said for something that's really a niche but I don't consider PyPy or musl to fall in that category. Users should be able to disable them easily which is the car (even more so if "skip=*muslinux*" is mentioned in release notes).  

> Hmm. Maybe we do an alpha i.e. v2.2.0a1 now so that eager beavers can try out musllinux now, but wait until #799 is ready for the proper release? I agree that it will be really useful to have #799 when we release musl, as that will really help the more complicated projects customise their configs, so they're more likely to adopt musllinux from the start.

Sounds like a good plan to me.

**Comment by joerick** [comment]
Alpha with musllinux support released! https://github.com/pypa/cibuildwheel/releases/tag/v2.2.0a1
